sdk/test-utils/recorder-new/src/core-v2-recorder.ts

```diff
@@ -36,15 +36,16 @@ import { Matcher, setMatcher } from "./matcher";
  * This client is meant for the core-v2 SDKs(depending on core-rest-pipeline) and
  * is supposed to be passed as an argument to the recorderHttpPolicy.
  */
-export class TestProxyHttpClient {
+export class RecorderClient {
```

Comment:
Part of the issue is that this can be passed as an `httpClient` through the parameters to other SDK clients.

I don’t think `recorderClient` is necessarily better than `Recorder` though. What I would like to see is the `httpClient` replacement to be something within the recorder. Perhaps using Will’s suggestion of `Recorder` as the name of this class, and then the http client as `recorder.httpClient`.

Reply:
@HarshaNalluru feel free to correct me if I'm wrong, but @sadasant my understanding is that we only allow passing it as an http client for corev1 pipelines (actually, this is _required_ for corev1), but it does not implement the corev2 HttpClient. There we just use a policy. This PR makes it so that the recorder does not directly implement HttpClient (v1), so that's why I'm suggesting what you're commenting on: rename this whole thing to just `Recorder`, eliminate the difference between v1 and v2 recorders, and give the recorder a `coreV1HttpClient` property to get the conforming implementation of HttpClient for that workflow.

Reply:
Naming it Recorder would conflict with the existing Recorder in the recorder package once I merge them.

I'll keep note of the rest of the points and think more.

Reply:
`rename this whole thing to just Recorder, eliminate the difference between v1 and v2 recorders, and give the recorder a coreV1HttpClient property to get the conforming implementation of HttpClient for that workflow.` @witemple-msft this sounds good to me! 🤔 was that a proposal? If so, I like it

Reply:
Jumping on the bandwagon here. I had a random thought last night but didn't end up commenting since I hadn't really fleshed it out. What if we didn't export the class but instead exported a factory function and interface, much like:

```ts
export interface Recorder {
  /* start(), stop(), etc */
}

interface CreateRecorderReturn {
  recorder: Recorder;
  httpClient: HttpClient;
  configureClient: (client: { pipeline: Pipeline }) => void
}

export function createRecorder(): CreateRecorderReturn {
  // obviously the internals of this function would change depending on the underlying design, but
  // something like this would work with the existing code (before this PR).
  const recorder = new TestProxyHttpClientCoreV1();
  return { recorder, httpClient: recorder, configureClient: recorder.configureClient };
}
```

So in your code you can do

```ts
const { recorder, configureClient } = createRecorder();
configureClient(/* your client */);

// and in your tests you run recorder.start(), recorder.stop(), etc.
```

and for the Core v1 case

```ts
const { recorder, httpClient } = createRecorder();
const client = new MyCoreV1Client(/* ... */, { httpClient });
```

From a user perspective this would totally decouple the recorder part and the part that instruments the client. And we can still change up the design internally to clean it up a little bit.

Reply:
> Naming it Recorder would conflict with the existing Recorder in the recorder package once I merge them.

We are planning to merge the old recorder into the new recorder package? That sounds like a recipe for confusion if the new recorder package exports something called just `Recorder` and that isn't the thing we're supposed to use.

Reply:
> We are planning to merge the old recorder into the new recorder package? 

moving the new recorder into the old recorder package.

> That sounds like a recipe for confusion if the new recorder package exports something called just Recorder and that isn't the thing we're supposed to use.
All good if it is called `RecorderClient`, right?

Reply:
> Jumping on the bandwagon here.

@timovv I can get behind that idea of having both the recorder and httpClient be destructured, but since we already have a touchpoint object in the form of the recorder, I think it might make more sense for `httpClient` to just be a property. Once packages move to corev2, they can forget it exists. Iterating a bit here, this is what I'm proposing more or less:

```typescript
function createHttpClientCoreV1(recorder: Recorder) {
  const client = new DefaultHttpClient();
  return {
    sendRequest: () => {/* implement sendRequest here using `client` and `recorder` */}
   }
}

/** Make a lazy value that can be deferred and only computed once. */
const once = <T>(make: () => T): () => T {
  let value;
  return () => value ??= make();
}

export class Recorder {
  // ... all the other Recorder implementation is still here
  private _httpClientCoreV1: () => HttpClient = once(() => createCoreV1HttpClient(this));
  
  get httpClientCoreV1(): HttpClient {
     return this._httpClientCoreV1();
  }
}
```

Then you use this by just accessing that HttpClient as a normal property:

```typescript
const client = new MyCoreV1Client(/* ... */, { httpClient: recorder.httpClientCoreV1 });
```

This is pretty lean and lets the design focus on the corev2 scenario. It would require only one core class for users to interact with, and this little `httpClientCoreV1` could be the escape-hatch to interoperate with corev1 clients, as opposed to having two separate recorder classes.

Reply:
> All good if it is called `RecorderClient`, right?

It's not world-ending, but I honestly don't think it's great. I didn't realize that squishing these packages together and merging the older codebase with the newer one was the plan. I like to think of an API as being the first line instruction manual about how to use a library and having both `Recorder`, which uses the old thing we don't want people to use anymore, and `RecorderClient` (which IMO sounds subordinate to `Recorder`, like the latter is a _component_ of the former) in the same API seems confusing to me.

We can always tell users/partners in the repo to use `RecorderClient`, deprecate `Recorder`, or whatever, but even easier than that is if there's only one thing that fits the bill in the package in the first place. I suppose if our goal is to get everyone onto the shared recorder and then _delete_ the old recorder code from the repo, then that would be fine, but then I'd argue for just keeping the old recorder in its own package until everyone migrates to the new one, and then just deleting it all-up. Then we could do a mechanical PR where we change the package names.

Basically, if you think just `Recorder` is a better name than `RecorderClient` on its face, then I think "we already used that name in the old package" isn't a good reason to toss it for a new name now. This is an internal tool, so we have a lot of options. If you think `RecorderClient` is a better name overall, then that's different.

Reply:
Your points make sense for the `Recorder` name.
But the old recorder is already published since it was needed for the arm packages to test in the auto rest repo.

If we go on the same route and force users to only use recorder-new, 
- all the exports such as env, isPlaybackMode, etc utils have to be re-exported from recorder-new
- we'll use the Recorder name
- we should also publish `recorder-new` (or come up with the sparse checkout plan asap as was discussed before)

Reply:
Ultimately I think anything that makes it so we don't have to publish the recorder is good, but I don't think that's the only option here. For example, all of the following seems like a realistic possibility:

You could delete the old recorder now and make recorder-new v2.0.0 of the recorder. Packages can keep getting v1.x from NPM until they migrate, since it's been published (do I understand this correctly?), then we just call this a major version upgrade (therefore changing version selector to `^2.0.0` is opting-in to recorder-new). You can even keep publishing the recorder if necessary this way (ultimately it'd be great if we could not do that and get a recorder artifact some other way for autorest CI.

> all the exports such as env, isPlaybackMode, etc utils have to be re-exported from recorder-new

Re-exported or copied outright, whichever you think is best. If recorder-new is v2 of recorder, then it might make sense to just copy the code.

Reply:
That actually sounds great.

Reply:
- RecorderClient -> Recorder done
- Followed mostly what @witemple-msft suggested
- `configureClientOptionsCoreV1` to be called on the options before passing them to the MyServiceClient